### PR TITLE
Refactor: redis cache

### PR DESCRIPTION
**Pull request by shimonewman**

### CODE REVIEW

File: phpunit.xml [hunk 1]
```diff
@@ -6,7 +6,7 @@
          convertNoticesToExceptions="true"
          convertWarningsToExceptions="true"
          processIsolation="false"
-         stopOnFailure="false"
+         stopOnFailure="true"
```
Comment (christyjacob4) [inline phpunit.xml:9]:
```suggestion
         stopOnFailure="false"
```

File: src/Cache/Adapter.php [hunk 2]
```diff
@@ -7,22 +7,31 @@ interface Adapter
     /**
      * @param  string  $key
      * @param  int  $ttl time in seconds
+     * @param  string  $hashKey optional
      * @return mixed
      */
-    public function load(string $key, int $ttl): mixed;
+    public function load(string $key, int $ttl, string $hashKey = ''): mixed;
```
Comment (christyjacob4) [inline src/Cache/Adapter.php:13]: Why not just `hash` ?

File: src/Cache/Adapter.php [hunk 3]
```diff
@@ -7,22 +7,31 @@ interface Adapter
...
+     */
+    public function list(string $key): array;
+
+    /**
+     * @param  string  $key
+     * @param  string  $hashKey optional
      * @return bool
      */
-    public function purge(string $key): bool;
+    public function purge(string $key, string $hashKey = ''): bool;
```
Comment (christyjacob4) [inline src/Cache/Adapter.php:34]: same as above. Why not just `hash` ?

File: tests/Cache/Base.php [hunk 4]
```diff
@@ -34,67 +34,68 @@ abstract class Base extends TestCase
...
 
         $this->assertEquals(false, $data);
     }
 
     public function testCaseInsensitivity(): void
     {
         // Ensure case in-sensitivity first (configured in adapter's setUp)
-        $data = self::$cache->save('planet', 'Earth');
+        $data = self::$cache->save('planet', 'Earth', 'planet');
+        var_dump($data);
```
Comment (christyjacob4) [inline tests/Cache/Base.php:73]:
```suggestion
```

File: src/Cache/Adapter.php (hunk 2 above)
Comment (shimonewman) [reply src/Cache/Adapter.php:13]: This is the terminology on the Redis interface

**Review by christyjacob4 — APPROVED** [review]

File: src/Cache/Adapter/Redis.php [hunk 5]
```diff
@@ -26,11 +27,13 @@ public function __construct(Client $redis)
     /**
      * @param  string  $key
      * @param  int  $ttl time in seconds
+     * @param  string  $hash optional
      * @return mixed
      */
-    public function load(string $key, int $ttl): mixed
+    public function load(string $key, int $ttl, string $hash = ''): mixed
```
Comment (fogelito) [inline src/Cache/Adapter/Redis.php:33]: Is $hash optional?

File: src/Cache/Adapter/Redis.php [hunk 6]
```diff
@@ -26,11 +27,13 @@ public function __construct(Client $redis)
...
     {
-        $redis_string = $this->redis->get($key);
+        $redis_string = $this->redis->hGet($key, $hash);
```
Comment (fogelito) [inline src/Cache/Adapter/Redis.php:35]: if $hash is empty to we want to use regular get?

File: src/Cache/Adapter/Redis.php [hunk 7]
```diff
@@ -47,34 +50,55 @@ public function load(string $key, int $ttl): mixed
...
         }
 
-        $cache = [
+        $value = json_encode([
             'time' => \time(),
             'data' => $data,
-        ];
+        ]);
+
+        if (! $value) {
```
Comment (fogelito) [inline src/Cache/Adapter/Redis.php:68]: You can remove this check if you add to json_encode this flag 
\json_encode([], flags: JSON_THROW_ON_ERROR);
and put it in the try catch block below

File: src/Cache/Adapter/Redis.php (hunk 6 above)
Comment (christyjacob4) [reply src/Cache/Adapter/Redis.php:35]: I dont think so. Regular GET is not very optimal when it comes to purging

File: src/Cache/Adapter/Redis.php (hunk 5 above)
Comment (shimonewman) [reply src/Cache/Adapter/Redis.php:33]: Yes

File: src/Cache/Adapter/Redis.php (hunk 7 above)
Comment (shimonewman) [reply src/Cache/Adapter/Redis.php]: Done

**Review by christyjacob4 — APPROVED** [review]

**Pull request merged**